File: solution/java/day20/src/main/java/domain/yahtzee/constrain/input/YahtzeeCalculator.java
package domain.yahtzee.constrain.input;

import io.vavr.Function1;

public class YahtzeeCalculator {
    public static int number(Roll roll, int number) {
        return calculate(r -> r.dice()
                        .filter(die -> die == number)
                        .sum()
                        .intValue()
                , roll);
    }

    private static int calculate(Function1<Roll, Integer> compute, Roll roll) {
        return compute.apply(roll);
    }

    private static int calculateNOfAKind(Roll roll, int n) {
        return calculate(r -> hasNOfAKind(r, n)
                        ? r.dice().sum().intValue()
                        : 0,
                roll);
    }

    public static int threeOfAKind(Roll roll) {
        return calculateNOfAKind(roll, 3);
    }

    public static int fourOfAKind(Roll roll) {
        return calculateNOfAKind(roll, 4);
    }

    public static int yahtzee(Roll roll) {
        return calculate(r -> hasNOfAKind(r, 5) ? Scores.YAHTZEE_SCORE : 0, roll);
    }

    public static int fullHouse(Roll roll) {
        return calculate(r -> {
            var dieFrequency = roll.groupDieByFrequency();
            return (dieFrequency.containsValue(3) && dieFrequency.containsValue(2)) ? Scores.HOUSE_SCORE : 0;
        }, roll);
    }

    public static int largeStraight(Roll roll) {
        return calculate(r -> r.dice()
                .sorted()
                .sliding(2)
                .forAll(pair -> pair.get(0) + 1 == pair.get(1)) ? Scores.LARGE_STRAIGHT_SCORE : 0, roll);
    }

    public static int smallStraight(Roll roll) {
        return calculate(r -> isSmallStraight(roll.toString()) ? 30 : 0, roll);
    }

    private static boolean isSmallStraight(String diceString) {
        return diceString.contains("1234") || diceString.contains("2345") || diceString.contains("3456");
    }

    private static boolean hasNOfAKind(Roll roll, int n) {
        return roll.groupDieByFrequency()
                .values()
                .exists(count -> count >= n);
    }

    public static int chance(Roll roll) {
        return calculate(r -> r.dice().sum().intValue(), roll);
    }

    private static class Scores {
        private static final int YAHTZEE_SCORE = 50;
        private static final int HOUSE_SCORE = 25;
        private static final int LARGE_STRAIGHT_SCORE = 40;
    }

    private YahtzeeCalculator() {
    }
}

File: solution/kotlin/day20/src/main/kotlin/domain/yahtzee/constrain/input/YahtzeeCalculator.kt
package domain.yahtzee.constrain.input

object YahtzeeCalculator {
    fun number(roll: Roll, number: Int): Int = calculate(
        { r -> r.dice.filter { die -> die == number }.sum() }, roll
    )

    fun threeOfAKind(roll: Roll): Int = calculateNOfAKind(roll, 3)
    fun fourOfAKind(roll: Roll): Int = calculateNOfAKind(roll, 4)
    fun yahtzee(roll: Roll): Int = calculate(
        { r -> if (hasNOfAKind(r, 5)) Scores.YAHTZEE_SCORE else 0 }, roll
    )

    private fun calculateNOfAKind(roll: Roll, n: Int): Int = calculate(
        { r -> if (hasNOfAKind(r, n)) r.dice.sum() else 0 }, roll
    )

    fun fullHouse(roll: Roll): Int = calculate(
        { r ->
            r.groupDieByFrequency()
                .let { dieFrequency ->
                    if (dieFrequency.containsValue(3) && dieFrequency.containsValue(2)) Scores.HOUSE_SCORE
                    else 0
                }
        }, roll
    )

    fun largeStraight(roll: Roll): Int = calculate(
        { r ->
            if (r.dice.sorted()
                    .windowed(2)
                    .all { pair -> pair[0] + 1 == pair[1] }
            ) Scores.LARGE_STRAIGHT_SCORE else 0
        }, roll
    )

    fun smallStraight(roll: Roll): Int = calculate({ r ->
        r.toSortedString()
            .let { diceString -> if (isSmallStraight(diceString)) 30 else 0 }
    }, roll)

    private fun isSmallStraight(diceString: String): Boolean =
        diceString.contains("1234") || diceString.contains("2345") || diceString.contains("3456")

    private fun hasNOfAKind(roll: Roll, n: Int): Boolean =
        roll.groupDieByFrequency()
            .values
            .any { count -> count >= n }

    fun chance(roll: Roll): Int = calculate(
        { r -> r.dice.sum() }, roll
    )

    private fun calculate(compute: (roll: Roll) -> Int, roll: Roll): Int = compute(roll)

    private object Scores {
        const val YAHTZEE_SCORE = 50
        const val HOUSE_SCORE = 25
        const val LARGE_STRAIGHT_SCORE = 40
    }
}